File: headless/src/main/org/nlogo/prim/_turtlevariabledouble.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Turtle;
import org.nlogo.api.AgentException;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Reporter;

public final strictfp class _turtlevariabledouble extends Reporter {
  public int vn;

  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax(Syntax.NumberType(), "-T--");
  }

  @Override
  public String toString() {
    if (world == null) {
      return super.toString() + ":" + vn;
    }
    return super.toString() + ":" + world.turtlesOwnNameAt(vn);
  }

  @Override
  public Object report(Context context) throws LogoException {
    try {
      return context.agent.getTurtleVariable(vn);
    } catch (AgentException ex) {
      throw new EngineException(context, this, ex.getMessage());
    }
  }

  public double report_1(Context context) {
    return ((Turtle) context.agent).getTurtleVariableDouble(vn);
  }

  public Double report_2(Context context) throws LogoException {
    try {
      return (Double) context.agent.getTurtleVariable(vn);
    } catch (AgentException ex) {
      throw new EngineException(context, this, ex.getMessage());
    }
  }
}

File: headless/src/main/org/nlogo/prim/_turtlevariabledouble.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.api.{ Syntax, AgentException }
import org.nlogo.nvm.{ Reporter, Context, EngineException }
import org.nlogo.agent.Turtle

class _turtlevariabledouble(private[this] var _vn: Int) extends Reporter {

  def this() = this(0)

  override def syntax =
    Syntax.reporterSyntax(Syntax.NumberType, "-T--")

  override def toString =
    super.toString + ":" +
      Option(world).map(_.turtlesOwnNameAt(vn)).getOrElse(vn.toString)

  // MethodRipper won't let us call a public method from report_1()
  // so we must keep vn and _vn separate - ST 9/22/12
  def vn = _vn
  def vn_=(vn: Int) { _vn = vn }

  override def report(context: Context) =
    report_2(context)

  def report_1(context: Context): Double =
    context.agent.asInstanceOf[Turtle].getTurtleVariableDouble(_vn)

  def report_2(context: Context): java.lang.Double =
    try context.agent.getTurtleVariable(_vn).asInstanceOf[java.lang.Double]
    catch { case ex: AgentException =>
      throw new EngineException(context, this, ex.getMessage) }

}
